File: Trees/450. Delete Node in a BST/solution.py

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root: return None

        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else: # key == root.val
            # if not root.left and not root.right: return None # redundant
            if not root.left: return root.right
            if not root.right: return root.left
            
            # find smallest node in right-subtree
            minNode = self.findMinNode(root.right)

            # delete smallest node and we'll swap smallest node with current root node afterwards
            root.right = self.deleteNode(root.right, minNode.val)

            # replace current root node with minNode of right subtree
            minNode.left = root.left
            minNode.right = root.right
            root = minNode
        return root
            
    def findMinNode(self, node):
        while node.left:
            node = node.left
        return node
    
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root: return None

        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else: # key == root.val
            # if not root.left and not root.right: return None # redundant
            if not root.left: return root.right
            if not root.right: return root.left
            
            # find smallest node in right-subtree
            minNode = self.findMinNode(root.right)
            
            # swap leaf node with current root node and delete leaf node
            root.val = minNode.val
            root.right = self.deleteNode(root.right, minNode.val)

        return root
            
    def findMinNode(self, node):
        while node.left:
            node = node.left
        return node

File: Trees/450. Delete Node in a BST/solution.cpp
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (!root) return nullptr;

        if (key < root->val) {
            root->left = deleteNode(root->left, key);
        } else if (key > root->val) {
            root->right = deleteNode(root->right, key);
        } else {
            // Node to be deleted found
            if (!root->left) return root->right;
            if (!root->right) return root->left;

            // Node with two children: get the inorder successor (smallest in the right subtree)
            TreeNode* successor = root->right;
            while (successor->left) {
                successor = successor->left;
            }
            // Copy the inorder successor's value to this node
            root->val = successor->val;
            // Delete the inorder successor
            root->right = deleteNode(root->right, successor->val);
        }
        return root;
    }
};